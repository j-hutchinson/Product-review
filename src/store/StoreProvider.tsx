import React, { createContext, useReducer } from 'react';
import { ActionType, StoreType } from '../types';
import reducer, { initialState } from './reducers';

interface Props {
    children: JSX.Element | JSX.Element[];
}

interface ContextType {
    state: StoreType;
    dispatch: (action: ActionType) => void;
}

export const StoreContext = createContext<ContextType>(null);

export const StoreProvider = ({ children }: Props): JSX.Element => {
    const [state, dispatch] = useReducer(reducer, initialState);

    return <StoreContext.Provider value={{ state, dispatch }}>{children}</StoreContext.Provider>;
};

