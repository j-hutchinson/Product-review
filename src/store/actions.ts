import { ActionType, Comment } from '../types';

const ADD_COMMENT = 'ADD_COMMENT';
const ADD_MULTIPLE_COMMENTS = 'ADD_MULTIPLE_COMMENTS';
const DELETE_COMMENTS = 'DELETE_COMMENTS';

const addComment = (comment: Comment): ActionType => {
	return { type: ADD_COMMENT, comment };
};

const addMultipleComments = (comments: Comment[]): ActionType => {
	return { type: ADD_MULTIPLE_COMMENTS, comments };
};

const deleteComments = (): ActionType => {
	return { type: DELETE_COMMENTS };
};

export {
	ADD_COMMENT,
	ADD_MULTIPLE_COMMENTS,
	DELETE_COMMENTS,
	addComment,
	addMultipleComments,
	deleteComments
};
