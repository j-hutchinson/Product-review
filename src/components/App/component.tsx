import React from 'react';
import styled from 'styled-components';
import { spacing } from '../../css/tokens';
import { mockData } from '../../data/mockData';
import { addMultipleComments, deleteComments } from '../../store/actions/actions';
import store from '../../store/store';
import AllComments from '../AllComments/component';
import PostComment from '../PostComment/component';
import Trends from '../Trends/component';

const StyledWrapper = styled.div`
  display: grid;
  grid-template-rows: max-content max-content;
  grid-gap: ${spacing.large};
  font-size: 14px;
  margin: 0;
  padding: ${spacing.medium};
  
  @media (min-width: 800px) {
    margin: ${spacing.medium} ${spacing.large};
    padding: ${spacing.xlarge};
  }
`;

const StyledSection = styled.div` 
  display: grid;
  grid-template-rows: 1fr 1fr;
  grid-template-columns: 1fr;
  grid-gap: ${spacing.medium};
  
  @media (min-width: 800px) {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: 1fr;
    grid-gap: ${spacing.medium};
  }
`;

const Button = styled.button`
  width: 100px;
  margin: 0 ${spacing.small};
`;

const StyledButtonWrapper = styled.div`
  display: flex;
`;

/* 
* Helper buttons for testing purposes
* Adds/Removes mock comments to store to test functionality
*/

const App = (): JSX.Element => (
  <StyledWrapper>
    <StyledSection>
      <PostComment />
      <Trends />
    </StyledSection>
    <AllComments />
    <StyledButtonWrapper>
      <Button className="multiple-button" onClick={() => store.dispatch(addMultipleComments(mockData))}>Use mock data</Button>
      <Button className="clear-button" onClick={() => store.dispatch(deleteComments())}>Clear</Button>
    </StyledButtonWrapper>
  </StyledWrapper >
);


export default App;
