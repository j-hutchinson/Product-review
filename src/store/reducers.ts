import { ADD_COMMENT, ADD_MULTIPLE_COMMENTS, DELETE_COMMENTS } from './actions';
import { ActionType, StoreType } from '../types';

export const initialState: StoreType = {
	allComments: []
};

const reducer = (state = initialState, action: ActionType): StoreType => {
	switch (action.type) {
		case ADD_COMMENT:
			return {
				allComments: [...state.allComments, action.comment]
			};
		case ADD_MULTIPLE_COMMENTS:
			return {
				allComments: [...state.allComments, ...action.comments]
			};
		case DELETE_COMMENTS:
			return {
				allComments: []
			};

		default:
			return state;
	}
};

export default reducer;
