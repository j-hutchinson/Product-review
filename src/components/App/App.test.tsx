import { shallow } from 'enzyme';
import React from 'react';
import store from '../../store/store';
import App from './component';

jest.mock('../../store/store', () => ({ dispatch: jest.fn(), }));

describe('App component', () => {
  afterEach(() => {
    jest.resetAllMocks();
  })

  test('component matches snapshot', () => {
    expect.assertions(1);

    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Multiple Button onClick prop triggers store update', () => {
    expect.assertions(2);

    const wrapper = shallow(<App />);
    wrapper.find('.multiple-button').props().onClick();

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      comments: expect.any(Object),
      type: "ADD_MULTIPLE_COMMENTS"
    });
  });

  test('Clear Button onClick prop triggers store update', () => {
    expect.assertions(2);

    const wrapper = shallow(<App />);
    wrapper.find('.clear-button').props().onClick();

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: "DELETE_COMMENTS"
    });
  });
});