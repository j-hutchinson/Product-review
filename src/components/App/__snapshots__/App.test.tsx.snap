// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component component matches snapshot 1`] = `
<styled.div>
  <styled.div>
    <PostComment />
    <Connect(Trends) />
  </styled.div>
  <Connect(AllComments) />
  <styled.div>
    <styled.button
      className="multiple-button"
      onClick={[Function]}
    >
      Use mock data
    </styled.button>
    <styled.button
      className="clear-button"
      onClick={[Function]}
    >
      Clear
    </styled.button>
  </styled.div>
</styled.div>
`;
